### DISCUSSION TITLE
Is there a roadmap for a 1.0 version of ktfmt?

### DISCUSSION THREAD

**Discussion by greyhairredbear** [facebookincubator/ktfmt#451] [General]
Is there such a thing? If so:

- What are the necessary features?
- What needs to be fixed?

**Comment by hick209** [comment]
@cgrushko, you might have context on this one here

**Comment by cgrushko** [comment]
ktfmt has been used internally at Meta for ~all code for more than 4 years now, so I think we can call it 1.0 today, if we'd like. 

The thing to pay attention to is semantic versioning - 
"
Given a version number MAJOR.MINOR.PATCH, increment the:

* MAJOR version when you make incompatible API changes
* MINOR version when you add functionality in a backward compatible manner
* PATCH version when you make backward compatible bug fixes
"

We may want to think about what 'API changes' mean in the context of ktfmt, which isn't intended to be used as a library (though it is, in practice, by gradle plugins for example). 

@strulovich  @hick209  @davidtorosyan  @cortinico  - am I missing something?

**Reply by davidtorosyan** [reply]
If we're doing semantic versioning, shouldn't it be 1.0.0?

Either way I'm not sure if it would be very meaningful to version the CLI with semantic versioning. What would even qualify as a backwards compatible change? Almost every change would either be a patch (fixing crashes) or major change (adjusting format). I guess adding CLI options would be minor. 

I'd say the library should get its own version in this case, if that's something we're prioritizing.

**Reply by greyhairredbear** [reply]
Thanks for the context to both of you, that is quite an interesting discussion. 

The points you brought up here make me think semantic versioning might not be a good fit for ktfmt. Apart from being interested in future plans for ktfmt, I probably brought this discussion up because `0.x` communicates something in a pre-release state. Which ktfmt obviously isn't. 

Two ideas here: 
- [CalVer](https://calver.org) with something like `year.minor` could be a better fit
- Tying ktfmt versioning to the kotlin compiler version would make sense. That maybe could imply  the (`major.minor-<ktfmt-version>`?) version of the language the can be formatted with ktfmt. There might be issues with this approach (would ktfmt need to maintain different language versions for an extended amount of time), but I think it makes a lot of sense, given ktfmt is a tool for formatting a programming language.

**Reply by cortinico** [reply]
> We may want to think about what 'API changes' mean in the context of ktfmt, which isn't intended to be used as a library (though it is, in practice, by gradle plugins for example).

Agree that this can practically be called a `1.0.0`

I believe we can stick to something like:
* MAJOR bumps - Changes to the Java API which are not backward compatible (i.e. we remove a function or a flag from the CLI).
* MINOR bumps - Changes that result in formatting changes (i.e. if you run 1.0.0 vs 1.1.0 your Kotlin code is likely to be reformatted differently)
* PATCH bumps - Bug fixes (i.e. internal crashes, anything else that won't directly manipulate the code).

Also we should look into using https://github.com/Kotlin/binary-compatibility-validator for API validation of major bumps

**Comment by hick209** [comment]
I like what @cortinico posted here and regarding how to handle our version numbers, I think we could follow that.

How about this proposal here for what to include when we do the 1.0.0 bump:
- Default style becomes Kotlin Lang (https://github.com/facebook/ktfmt/discussions/446)
- Rename Facebook style to Meta style (https://github.com/facebook/ktfmt/discussions/446)
- Turn on trailing commas on Meta (https://github.com/facebook/ktfmt/issues/442)
- Remove Dropbox style (as this is unsupported according to @cgrushko, [source](https://github.com/facebook/ktfmt/discussions/172#discussioncomment-393145), https://github.com/facebook/ktfmt/discussions/172)

Thoughts?
Anything that should be included in this list or removed from it?

If no major complaints about this plan are presented in a week, I'll start putting this plan into action.

**Reply by ColtonIdle** [reply]
I still find it weird to call a style by the name of the company. Esp if it only has to do with spacing. Have we explored other options there?

**Reply by hick209** [reply]
Not really. Those are indeed the styles used by those Meta and Google, and they were created for that therefore those are the names.
We created Kotlinlang because it was slightly different to Dropbox style and something people were requesting for (unless there's more context here that I'm missing cc @cgrushko).

Other than you finding it "weird" do you see some other reason why we should try something different? Any proposals?

**Reply by ColtonIdle** [reply]
I guess just having an explanation of each, or a link that goes to each "style" would be nice. I currently have been using ktfmt for years and I don't know the difference except that one of them is 4 spaces vs 2 spaces. so i thought just having it named by spaces would make more sense.
but if kotlinlang actually follows: https://kotlinlang.org/docs/coding-conventions.html
and google follows: https://google.github.io/styleguide/javaguide.html
then maybe adding an android style as well (esp since androidx team is now using ktfmt): https://developer.android.com/kotlin/style-guide
is there a public styleguide for dropbox?

**Reply by hick209** [reply]
We're deleting Dropbox one (maybe today) as it's the same as kotlinlang.

You have a fair point about the documentation here. Luckily we have @grodin that is currently helping us out on writing better documentation for the styles (with a proper `--help` option #477), which I would hope we can have landed by 0.51 which hopefully will be the last one before 1.0.0.

Now regarding Android X style, that'd be up to them (Google folks) to add here, which I don't know if they plan to or not, cc @nreid260, @liutikas, @omarismail94 in case you can add some color in regards to this.

**Reply by omarismail94** [reply]
For AndroidX, we use [kotlinglang style](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:buildSrc/private/src/main/kotlin/androidx/build/Ktfmt.kt;l=150). There are multiple teams at Google using `ktfmt`, but not all of them are using kotlinlang style.

**Reply by ColtonIdle** [reply]
cool. yeah i kinda dont care if the names are "company" names as long as its clear what the style is representing.

like i said before. ive used ktfmt for years and i thought the only difference between the names for 2 vs 4 spaces. lol

**Reply by hick209** [reply]
@omarismail94 interesting. I'll pull you into the conversation here then https://github.com/facebook/ktfmt/issues/442

**Reply by nreid260** [reply]
`--google-style` essentially represents "the style that Kotlin-at-Google team enforces internally, which might not be the same as styles published by other Google affiliated entities (e.g. Android stuff)". We also use it to flag-guard features that other styles might not have agreed to (e.g. the trailing commas).
